File: TRViS.UITests.Apple/Sources/Shared/OriginalTimetableV6PageObject.swift
// OriginalTimetableV6PageObject.swift
// Page object for the V6 (Bold Editorial) Original Timetable page.
// Mirrors TRViS.UITests/Pages/OriginalTimetableV6PageObject.cs.

import XCTest

class OriginalTimetableV6PageObject {

    private let app: XCUIApplication
    private let base: BaseUITestCase

    init(app: XCUIApplication, base: BaseUITestCase) {
        self.app = app
        self.base = base
    }

    /// Waits up to `timeout` for V6 to reach a rendered state. Masthead,
    /// CurrentBlock, or EmptyState (tablet or compact mirror) satisfies the
    /// wait.
    @discardableResult
    func waitForRendered(timeout: TimeInterval = 30) -> Bool {
        let anchors = [
            AutomationIds.OriginalTimetable.V6.masthead,
            AutomationIds.OriginalTimetable.V6.currentBlock,
            AutomationIds.OriginalTimetable.V6.emptyState,
            AutomationIds.OriginalTimetable.V6.compactMasthead,
            AutomationIds.OriginalTimetable.V6.compactCurrentBlock,
            AutomationIds.OriginalTimetable.V6.compactEmptyState,
        ]
        let deadline = Date().addingTimeInterval(timeout)
        while Date() < deadline {
            for id in anchors {
                if let _ = base.waitForElement(id: id, timeout: 0.5) {
                    return true
                }
            }
            Thread.sleep(forTimeInterval: 0.2)
        }
        return false
    }

    /// True when any V6 EmptyState Label (tablet or compact) is visible.
    func isEmptyStateVisible(timeout: TimeInterval = 5) -> Bool {
        if base.waitForElement(
            id: AutomationIds.OriginalTimetable.V6.emptyState, timeout: timeout
        ) != nil { return true }
        return base.waitForElement(
            id: AutomationIds.OriginalTimetable.V6.compactEmptyState, timeout: timeout
        ) != nil
    }

    /// True when the tablet CurrentBlock is visible.
    func isCurrentBlockVisible(timeout: TimeInterval = 5) -> Bool {
        return base.waitForElement(
            id: AutomationIds.OriginalTimetable.V6.currentBlock, timeout: timeout
        ) != nil
    }

    /// Reads the CurrentBlock's StationName Label, empty when not in the tree.
    func getCurrentStation() -> String {
        guard let el = base.waitForElement(
            id: AutomationIds.OriginalTimetable.V6.currentBlockStationName, timeout: 10
        ) else { return "" }
        let label = el.label
        if !label.isEmpty { return label }
        if let v = el.value as? String { return v }
        return ""
    }

    /// True when ANY MarkerBadge (V6-prefixed row OR CurrentBlock OR Compact)
    /// is visible. Used by the negative pre-condition test.
    func isAnyMarkerBadgeVisible(timeout: TimeInterval = 5) -> Bool {
        let deadline = Date().addingTimeInterval(timeout)
        let cbId = AutomationIds.OriginalTimetable.V6.currentBlockMarkerBadge
        let cbCompactId = AutomationIds.OriginalTimetable.V6.compactCurrentBlockMarkerBadge
        while Date() < deadline {
            let predicate = NSPredicate(
                format: "(identifier BEGINSWITH 'OriginalTimetable.V6.Row.' OR identifier == %@ OR identifier == %@) AND identifier ENDSWITH '.MarkerBadge'",
                cbId, cbCompactId
            )
            let query = app.descendants(matching: .any).matching(predicate)
            for i in 0..<min(query.count, 10) {
                let el = query.element(boundBy: i)
                if el.exists {
                    let frame = el.frame
                    if frame.size.width > 0 && frame.size.height > 0 {
                        return true
                    }
                }
            }
            Thread.sleep(forTimeInterval: 0.15)
        }
        return false
    }
}

File: TRViS.UITests/Pages/OriginalTimetableV6PageObject.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Appium;
using TRViS.UITests.Infrastructure;

namespace TRViS.UITests.Pages;

/// <summary>
/// Page object for OriginalTimetableV6Page ("Bold Editorial"). Masthead +
/// TrainStripe + PastChips + CurrentBlock + UpcomingList. Both tablet
/// (&gt;=600pt) and compact layouts implemented. The page is "rendered" when
/// the Masthead, CurrentBlock, EmptyState, or their compact mirrors are
/// reachable.
/// </summary>
public class OriginalTimetableV6PageObject : PageObject
{
	public OriginalTimetableV6PageObject(AppiumDriver driver) : base(driver) { }

	public AppiumElement Masthead
		=> WaitForElement(AutomationIds.OriginalTimetable.V6.Masthead, TimeSpan.FromSeconds(30));
	public AppiumElement CurrentBlock
		=> WaitForElement(AutomationIds.OriginalTimetable.V6.CurrentBlock, TimeSpan.FromSeconds(30));
	public AppiumElement CurrentStation
		=> WaitForElement(AutomationIds.OriginalTimetable.V6.CurrentBlockStationName, TimeSpan.FromSeconds(30));
	public AppiumElement EmptyState
		=> WaitForElement(AutomationIds.OriginalTimetable.V6.EmptyState, TimeSpan.FromSeconds(30));

	/// <summary>
	/// Waits until V6 has finished its first render. Any of the masthead /
	/// current-block / empty-state anchors (tablet or compact) is sufficient
	/// evidence of paint.
	/// </summary>
	public bool WaitForRendered(double timeoutSeconds = 30)
	{
		var deadline = DateTime.UtcNow.AddSeconds(timeoutSeconds);
		Driver.Manage().Timeouts().ImplicitWait = TimeSpan.Zero;
		try
		{
			string[] anchors = new[]
			{
				AutomationIds.OriginalTimetable.V6.Root,
				AutomationIds.OriginalTimetable.V6.Masthead,
				AutomationIds.OriginalTimetable.V6.CurrentBlock,
				AutomationIds.OriginalTimetable.V6.EmptyState,
				AutomationIds.OriginalTimetable.V6.CompactMasthead,
				AutomationIds.OriginalTimetable.V6.CompactCurrentBlock,
				AutomationIds.OriginalTimetable.V6.CompactEmptyState,
			};
			while (DateTime.UtcNow < deadline)
			{
				foreach (var id in anchors)
				{
					try
					{
						if (Driver.FindElements(AutomationIdLocator(id)).Count > 0)
							return true;
					}
					catch { }
				}
				Thread.Sleep(200);
			}
			return false;
		}
		finally
		{
			Driver.Manage().Timeouts().ImplicitWait = TimeSpan.FromSeconds(10);
		}
	}

	/// <summary>True when an EmptyState placeholder is visible (tablet OR compact).</summary>
	public bool IsEmptyStateVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V6.EmptyState, timeoutSeconds)
			|| PollDisplayed(AutomationIds.OriginalTimetable.V6.CompactEmptyState, timeoutSeconds);

	/// <summary>True when the tablet CurrentBlock is visible (active-train state).</summary>
	public bool IsCurrentBlockVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V6.CurrentBlock, timeoutSeconds);

	/// <summary>True when the tablet layout grid is in the tree.</summary>
	public bool IsTabletLayoutPresent(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V6.TabletGrid, timeoutSeconds);

	/// <summary>True when the compact layout grid is in the tree.</summary>
	public bool IsCompactLayoutPresent(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V6.CompactGrid, timeoutSeconds);

	/// <summary>
	/// Reads the current-block's station-name label. Empty when the label is
	/// not in the tree (e.g. no active train).
	/// </summary>
	public string GetCurrentStation()
	{
		try
		{
			return CurrentStation.Text ?? string.Empty;
		}
		catch (NoSuchElementException)
		{
			return string.Empty;
		}
	}

	/// <summary>True when the CurrentBlock's MarkerBadge is currently visible (tablet layout).</summary>
	public bool IsCurrentBlockMarkerBadgeVisible(double timeoutSeconds = 3)
		=> PollDisplayed(AutomationIds.OriginalTimetable.V6.CurrentBlockMarkerBadge, timeoutSeconds);
}
